Engine/Component/ComponentParticleSystem.cpp

```diff
@@ -0,0 +1,342 @@
+				else
+				{
+					billboard->Render(p.position);
+				}
+				
+			}
+		}
+	}
+	
+	glDisable(GL_BLEND);
+}
+void ComponentParticleSystem::UpdateParticle(Particle& p)
+{
+	float time_spend = p.time_passed;
+	p.life -= App->time->real_time_delta_time; // reduce life
+	time_spend -= p.life;
+	p.position += p.velocity * App->time->real_time_delta_time;
+
+	//alpha fade
+
+	if (fade)
+	{
+		p.color.w -= App->time->real_time_delta_time * (fade_time / 1000);
+		billboard->color[3] = p.color.w;
+	}
+	else
+	{
+		billboard->color[3] = 1.0F;
+	}
+
+	//fade color
+	if (fade_between_colors)
+	{
+		float time = (time_spend / 1000) * (color_fade_time / 100);
+		float temp_color[3] = { p.color.x ,p.color.y ,p.color.z };
+		p.color.x = (1 - time) * p.color.x + time * color_to_fade[0];
+		p.color.y = (1 - time) * p.color.y + time * color_to_fade[1];
+		p.color.z = (1 - time) * p.color.z + time * color_to_fade[2];
+	}
+	billboard->color[0] = p.color.x;
+	billboard->color[1] = p.color.y;
+	billboard->color[2] = p.color.z;
+
+	//size
+	billboard->width = particles_width * p.particle_scale;
+	billboard->height = particles_height * p.particle_scale;
+}
+void ComponentParticleSystem::SetParticleTexture(uint32_t texture_uuid)
+{
+	this->texture_uuid = texture_uuid;
+	billboard->ChangeTexture(texture_uuid);
+}
+void ComponentParticleSystem::Delete()
+{
+	App->renderer->RemoveComponentParticleSystem(this);
+}
+
+void ComponentParticleSystem::SpecializedSave(Config& config) const
+{
+
+	config.AddInt(static_cast<int>(type_of_particle_system), "Type of particle system");
+	config.AddBool(loop, "Loop");
+	config.AddInt(max_particles, "Max Particles");
+	config.AddInt(last_used_particle, "Last used particle");
+	config.AddInt(nr_new_particles, "Number of new particles");
+	config.AddBool(active, "Active");
+	config.AddInt(min_size_of_particle, "Max Size Particles");
+	config.AddInt(max_size_of_particle, "Min Size Particles");
+	config.AddFloat(particles_width, "Particle Width");
+	config.AddFloat(particles_height, "Particle Height");
+	config.AddBool(size_random, "Size random");
+
+	config.AddFloat(velocity_particles, "Velocity of particles");
+
+	config.AddFloat(time_counter, "Time Counter");
+	config.AddFloat(time_between_particles, "Time Between Particles");
+	config.AddFloat(particles_life_time, "Particles Life Time");
+
+	config.AddBool(follow_owner, "Follow owner");
+
+	config.AddBool(enabled_random_x, "Random X position");
+	config.AddInt(max_range_random_x, "Max range position x");
+	config.AddInt(min_range_random_x, "Min range position x");
+	config.AddInt(position_x, "Position X");
+
+	config.AddBool(enabled_random_z, "Random Z position");
+	config.AddInt(max_range_random_z, "Max range position z");
+	config.AddInt(min_range_random_z, "Min range position z");
+	config.AddInt(position_z, "Position Z");
+
+	config.AddFloat(inner_radius, "Inner Radius");
+	config.AddFloat(outer_radius, "Outer Radius");
+
+	config.AddFloat(color_particle[0], "Color Particle R");
```

Comment:
Wouldn't it be easy to use a float4?

Reply:
Is the same using float4 and float [4]

Reply:
I think you could save on the JSON one value of float4 instead of saving each value separately